analysis/topEFT/datacard_maker.py

```diff
@@ -215,13 +231,9 @@ def export2d(h):
             pname = self.rename[p]+'_' if p in self.rename else p
             pname.replace('_4F','').replace('_ext','')
             if 'njet' in variable:
-                if   '2l' in channel: h_base = h_base.rebin('njets', hist.Bin("njets",  "Jet multiplicity ", [4,5,6,7]))
-                elif '3l' in channel: h_base = h_base.rebin('njets', hist.Bin("njets",  "Jet multiplicity ", [2,3,4,5]))
-                elif '4l' in channel: h_base = h_base.rebin('njets', hist.Bin("njets",  "Jet multiplicity ", [2,3,4]))
-            if 'ht' in variable:
-                h_base = h_base.rebin('ht', hist.Bin("ht", "H$_{T}$ (GeV)", [0, 100, 200, 300, 400, 2000]))
-            if 'ptbl' in variable:
-                h_base = h_base.rebin('ptbl', hist.Bin("ptbl", "$p_{T}^{b\mathrm{-}jet+\ell_{min(dR)}}$", [0, 50, 100, 200, 400, 2000]))
+                h_base = h_base.rebin(variable, hist.Bin(variable,  h.axis(variable).label, self.analysis_bins[variable][channel[:2]]))
+            else:
+                h_base = h_base.rebin(variable, hist.Bin(variable,  h.axis(variable).label, self.analysis_bins[variable]))
```

Comment:
1. The `channel[:2]` slicing feels really fragile, but I see that this sort of thing is used elsewhere. Just makes it very reliant on the channel naming following a specific form.
2. I didn't notice earlier when you defined `self.analysis_bins`, but the fact that one set of keys has a `list` as value and the other has a `dict` as the value makes me think that over specializing this dictionary. I'm guessing the reasoning here was b/c `ptbl` and `ht` are meant to have the same binning for all channels, but what if we want to change that? or what we had some other distribution that we want to have different bins for different channels? My suggestion here would be one of two approaches. The first would be to change `self.analysis_bins` to _always_ have the value be a dictionary and if you don't want to have to specify the same exact list for every channel of `ptbl` and `ht`, just add a `default` key that is used if `channel[:2]` can't be found. The other approach would be to instead change the `if 'njet' in variable` check to instead be `if isinstance(self.analysis_bins[variable],dict)`. This way you're basically saying that if we specify a list that means the binning is the same for all channels, otherwise it is channel specific. I'm personally in favor of the 2nd method since it keeps the original dictionary structure as is and simultaneously removes the hard-coded dependence on the `'njet'` string.

Reply:
I also like option 2, and I switched to a regex to find the jet bin number.